File: app/src/main/java/com/example/khughes/machewidget/CarStatusWidget_2x5.java
package com.example.khughes.machewidget;

import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.view.View;
import android.widget.RemoteViews;

import androidx.preference.PreferenceManager;

import com.example.khughes.machewidget.CarStatus.CarStatus;

import java.util.ArrayList;
import java.util.Map;

/**
 * Implementation of App Widget functionality.
 */
public class CarStatusWidget_2x5 extends CarStatusWidget {

    // Define actions for clicking on various icons, including the widget itself
    protected void setCallbacks(Context context, RemoteViews views, int id) {
        views.setOnClickPendingIntent(R.id.wireframe, getPendingSelfIntent(context, id, PROFILE_CLICK));

        boolean showAppLinks = PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(context.getResources().getString(R.string.show_app_links_key), true);
        views.setOnClickPendingIntent(R.id.leftappbutton, getPendingSelfIntent(context, id, showAppLinks ? LEFT_BUTTON_CLICK : WIDGET_CLICK));
        views.setOnClickPendingIntent(R.id.rightappbutton, getPendingSelfIntent(context, id, showAppLinks ? RIGHT_BUTTON_CLICK : WIDGET_CLICK));

        super.setCallbacks(context, views, id);
    }

    protected void drawVehicleImage(Context context, RemoteViews views, CarStatus carStatus, int vehicleColor, ArrayList<Integer> whatsOpen, Map<String, Integer> vehicleImages) {
        whatsOpen = new ArrayList<>();
        whatsOpen.add(isWindowClosed(carStatus.getDriverWindow()) ? null : vehicleImages.get(Vehicle.LEFT_FRONT_WINDOW));
        whatsOpen.add(isWindowClosed(carStatus.getPassengerWindow()) ? null : vehicleImages.get(Vehicle.RIGHT_FRONT_WINDOW));
        whatsOpen.add(isWindowClosed(carStatus.getLeftRearWindow()) ? null : vehicleImages.get(Vehicle.LEFT_REAR_WINDOW));
        whatsOpen.add(isWindowClosed(carStatus.getRightRearWindow()) ? null : vehicleImages.get(Vehicle.RIGHT_REAR_WINDOW));
        super.drawVehicleImage(context, views, carStatus, vehicleColor, whatsOpen, vehicleImages);
    }

    private void updateAppWidget(Context context, AppWidgetManager appWidgetManager,
                                 int appWidgetId, InfoRepository info) {

        // Construct the RemoteViews object
        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_2x5);

        // Setup actions for specific widgets
        setCallbacks(context, views, appWidgetId);

        // Set background transparency
        setBackground(context, views);

        // Find which user is active.
        UserInfo userInfo = info.getUser();
        if (userInfo == null) {
            return;
        }

        // Find the vehicle for this widget
        VehicleInfo vehicleInfo = getVehicleInfo(context, info, appWidgetId);
        if (vehicleInfo == null) {
            return;
        }

        views.setTextViewText(R.id.profile, vehicleInfo.getNickname());
//        views.setTextViewText(R.id.profile, "My Mach-E");

        // Get conversion factors for Metric vs Imperial measurement units
        int units = Integer.parseInt(PreferenceManager.getDefaultSharedPreferences(context)
                .getString(
                        context.getResources().getString(R.string.units_key),
                        context.getResources().getString(R.string.units_system)
                ));
        double distanceConversion;
        String distanceUnits;
        if ((units == Constants.UNITS_SYSTEM && userInfo.getUomSpeed() != null && userInfo.getUomSpeed().equals("MPH")) || units == Constants.UNITS_IMPERIAL) {
            distanceConversion = Constants.KMTOMILES;
            distanceUnits = "miles";
        } else {
            distanceConversion = 1.0;
            distanceUnits = "km";
        }
        double pressureConversion;
        String pressureUnits;
        if ((units == Constants.UNITS_SYSTEM && userInfo.getUomPressure() != null && userInfo.getUomPressure().equals("PSI")) || units == Constants.UNITS_IMPERIAL) {
            pressureConversion = Constants.KPATOPSI;
            pressureUnits = "psi";
        } else if (units == Constants.UNITS_SYSTEM && userInfo.getUomPressure() != null && userInfo.getUomPressure().equals("BAR")) {
            pressureConversion = Constants.KPATOBAR;
            pressureUnits = "bar";
        } else {
            pressureConversion = 1.0;
            pressureUnits = "kPa";
        }

        CarStatus carStatus = vehicleInfo.getCarStatus();
        if (carStatus == null || carStatus.getVehiclestatus() == null) {
            views.setTextViewText(R.id.lastRefresh, "Unable to retrieve status information.");
            appWidgetManager.updateAppWidget(appWidgetId, views);
            return;
        }

        boolean isICEOrHybrid = carStatus.isPropulsionICEOrHybrid(carStatus.getPropulsion());
        boolean isPHEV = carStatus.isPropulsionPHEV(carStatus.getPropulsion());

        views.setViewVisibility(R.id.lock_gasoline, isICEOrHybrid ? View.VISIBLE : View.GONE);
        views.setViewVisibility(R.id.bottom_gasoline, isICEOrHybrid ? View.VISIBLE : View.GONE);
        views.setViewVisibility(R.id.lock_electric, isICEOrHybrid ? View.GONE : View.VISIBLE);
        views.setViewVisibility(R.id.bottom_electric, isICEOrHybrid ? View.GONE : View.VISIBLE);
        views.setViewVisibility(R.id.plug, isICEOrHybrid ? View.GONE : View.VISIBLE);
        setPHEVCallbacks(context, views, isPHEV, appWidgetId, "showGasoline");

        // Show last refresh, odometer, OTA status
        String timeFormat = userInfo.getCountry().equals("USA") ? Constants.LOCALTIMEFORMATUS : Constants.LOCALTIMEFORMAT;
        drawLastRefresh(context, views, carStatus, timeFormat);
        drawOdometer(views, carStatus, distanceConversion, distanceUnits);
        drawOTAInfo(context, views, vehicleInfo, timeFormat);

        // Location
        if (PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(context.getResources().getString(R.string.show_location_key), true)) {
            views.setViewVisibility(R.id.location_container, View.VISIBLE);
            updateLocation(context, views, carStatus.getLatitude(), carStatus.getLongitude());
        } else {
            views.setViewVisibility(R.id.location_container, View.GONE);
        }

        // Ignition, alarm/sleep, plug icons
        drawIcons(views, carStatus);

        // Draw range and fuel/gas stuff
        boolean twoLines = false;
        drawRangeFuel(context, views, carStatus, info, vehicleInfo,
                distanceConversion, distanceUnits, twoLines);

        // Tire pressures
        updateTire(context, views, carStatus.getLeftFrontTirePressure(), carStatus.getLeftFrontTireStatus(),
                pressureUnits, pressureConversion, R.id.lt_ft_tire);
        updateTire(context, views, carStatus.getRightFrontTirePressure(), carStatus.getRightFrontTireStatus(),
                pressureUnits, pressureConversion, R.id.rt_ft_tire);
        updateTire(context, views, carStatus.getLeftRearTirePressure(), carStatus.getLeftRearTireStatus(),
                pressureUnits, pressureConversion, R.id.lt_rr_tire);
        updateTire(context, views, carStatus.getRightRearTirePressure(), carStatus.getRightRearTireStatus(),
                pressureUnits, pressureConversion, R.id.rt_rr_tire);

        // Get the right images to use for this vehicle
        Map<String, Integer> vehicleImages = Vehicle.getVehicle(vehicleInfo.getVIN()).getHorizontalDrawables();

        // See if we should guess vehicle color
        if (VehicleColor.scanImageForColor(context, vehicleInfo)) {
            info.setVehicle(vehicleInfo);
        }

        // If vehicle is a Mach-E First Edition, show mirrors in body color
        if(VehicleColor.isFirstEdition(context, vehicleInfo.getVIN())) {
            vehicleImages.put(Vehicle.BODY_SECONDARY, R.drawable.mache_secondary_no_mirrors_horz);
        }

        // Draw the vehicle image
        drawVehicleImage(context, views, carStatus, vehicleInfo.getColorValue(), null, vehicleImages);

        updateLinkedApps(context, views);

        // Instruct the widget manager to update the widget
        appWidgetManager.updateAppWidget(appWidgetId, views);
    }

    @Override
    public void onAppWidgetOptionsChanged(Context context, AppWidgetManager appWidgetManager, int appWidgetId, Bundle newOptions) {
        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_2x5);
        appWidgetManager.partiallyUpdateAppWidget(appWidgetId, views);
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        final InfoRepository[] info = {null};

        Handler handler = new Handler(Looper.getMainLooper()) {
            @Override
            public void handleMessage(Message msg) {
                UserInfo user = info[0].getUser();
                if (user == null) {
                    LogFile.d(context, MainActivity.CHANNEL_ID, "CarStatusWidget_2x5.onUpdate(): no userinfo found");
                    return;
                }

                // There may be multiple widgets active, so update all of them
                for (int appWidgetId : appWidgetIds) {
                    updateAppWidget(context, appWidgetManager, appWidgetId, info[0]);
                }
            }
        };

        new Thread(() -> {
            info[0] = new InfoRepository(context);
            handler.sendEmptyMessage(0);
        }).start();

    }

    @Override
    public void onEnabled(Context context) {
        // Enter relevant functionality for when the first widget is created
    }

    @Override
    public void onDisabled(Context context) {
        // Enter relevant functionality for when the last widget is disabled
    }

    public void onReceive(Context context, Intent intent) {
        String action = intent.getAction();

        if (action.equals(PHEVTOGGLE_CLICK)) {
            String mode = intent.getStringExtra("nextMode");
            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_2x5);
            String nextMode;
            if (mode.equals("showGasoline")) {
                nextMode = "showElectric";
                views.setViewVisibility(R.id.bottom_electric, View.GONE);
                views.setViewVisibility(R.id.bottom_gasoline, View.VISIBLE);
            } else {
                nextMode = "showGasoline";
                views.setViewVisibility(R.id.bottom_electric, View.VISIBLE);
                views.setViewVisibility(R.id.bottom_gasoline, View.GONE);
            }
            int appWidgetId = intent.getIntExtra(APPWIDGETID, -1);
            setPHEVCallbacks(context, views, true, appWidgetId, nextMode);
            appWidgetManager.partiallyUpdateAppWidget(appWidgetId, views);
        } else {
            super.onReceive(context, intent);
        }
    }
}

File: app/src/main/java/com/example/khughes/machewidget/CarStatusWidget_2x5.kt
package com.example.khughes.machewidget

import com.example.khughes.machewidget.Vehicle.Companion.getVehicle
import com.example.khughes.machewidget.VehicleColor.Companion.scanImageForColor
import android.widget.RemoteViews
import android.appwidget.AppWidgetManager
import android.content.Context
import android.os.Bundle
import android.content.Intent
import android.view.View
import androidx.preference.PreferenceManager
import kotlinx.coroutines.*

/**
 * Implementation of App Widget functionality.
 */
class CarStatusWidget_2x5 : CarStatusWidget() {

    // Define actions for clicking on various icons, including the widget itself
    override fun setCallbacks(context: Context, views: RemoteViews, id: Int) {
        views.setOnClickPendingIntent(
            R.id.wireframe,
            getPendingSelfIntent(context, id, PROFILE_CLICK)
        )
        val showAppLinks = PreferenceManager.getDefaultSharedPreferences(context)
            .getBoolean(context.resources.getString(R.string.show_app_links_key), true)
        views.setOnClickPendingIntent(
            R.id.leftappbutton,
            getPendingSelfIntent(context, id, if (showAppLinks) LEFT_BUTTON_CLICK else WIDGET_CLICK)
        )
        views.setOnClickPendingIntent(
            R.id.rightappbutton,
            getPendingSelfIntent(
                context,
                id,
                if (showAppLinks) RIGHT_BUTTON_CLICK else WIDGET_CLICK
            )
        )
        super.setCallbacks(context, views, id)
    }

    override fun drawVehicleImage(
        context: Context,
        views: RemoteViews,
        carStatus: CarStatus,
        vehicleColor: Int,
        whatzOpen: MutableList<Int>?,
        vehicleImages: Map<String, Int>
    ) {
        val whatsOpen = whatzOpen ?: mutableListOf()

//        if (!isWindowClosed(carStatus.vehicle.vehicleStatus.windowPosition?.driverWindowPosition?.value)) {
//            whatsOpen.add(vehicleImages[Vehicle.LEFT_FRONT_WINDOW]!!)
//        }
//        if (!isWindowClosed(carStatus.vehicle.vehicleStatus.windowPosition?.passWindowPosition?.value)) {
//            whatsOpen.add(vehicleImages[Vehicle.RIGHT_FRONT_WINDOW]!!)
//        }
//        if (!isWindowClosed(carStatus.vehicle.vehicleStatus.windowPosition?.rearDriverWindowPos?.value)) {
//            whatsOpen.add(vehicleImages[Vehicle.LEFT_REAR_WINDOW]!!)
//        }
//        if (!isWindowClosed(carStatus.vehicle.vehicleStatus.windowPosition?.rearPassWindowPos?.value)) {
//            whatsOpen.add(vehicleImages[Vehicle.RIGHT_REAR_WINDOW]!!)
//        }
        super.drawVehicleImage(context, views, carStatus, vehicleColor, whatsOpen, vehicleImages)
    }

    private fun updateAppWidget(
        context: Context, appWidgetManager: AppWidgetManager,
        appWidgetId: Int, info: InfoRepository?
    ) {

        // Construct the RemoteViews object
        val views = RemoteViews(context.packageName, R.layout.widget_2x5)

        // Setup actions for specific widgets
        setCallbacks(context, views, appWidgetId)

        // Set background transparency
        setBackground(context, views)

        // Find which user is active.

        // Find the vehicle for this widget
        val vehicleInfo = getVehicleInfo(context, info!!, appWidgetId) ?: return
        views.setTextViewText(R.id.profile, vehicleInfo.carStatus.vehicle.nickName)
        //        views.setTextViewText(R.id.profile, "My Mach-E");

        // Get conversion factors and descriptions for measurement units
        val units = PreferenceManager.getDefaultSharedPreferences(context)
            .getString(
                context.resources.getString(R.string.units_key),
                context.resources.getString(R.string.units_mphpsi)
            )!!.toInt()
        val distanceConversion: Double
        val distanceUnits: String
        if (units == Constants.UNITS_MPHPSI) {
            distanceConversion = Constants.KMTOMILES
            distanceUnits = "miles"
        } else {
            distanceConversion = 1.0
            distanceUnits = "km"
        }
        val pressureConversion: Double
        val pressureUnits: String
        when (units) {
            Constants.UNITS_KPHPSI, Constants.UNITS_MPHPSI -> {
                pressureConversion = Constants.KPATOPSI
                pressureUnits = "psi"
            }
            Constants.UNITS_KPHBAR -> {
                pressureConversion = Constants.KPATOBAR
                pressureUnits = "bar"
            }
            else -> {
                pressureConversion = 1.0
                pressureUnits = "kPa"
            }
        }

        val carStatus = vehicleInfo.carStatus
        if(carStatus.vehicle.vehicleId == "") {
            return
        }

//        if (carStatus == null || carStatus.vehicle.vehicleStatus == null) {
//            views.setTextViewText(R.id.lastRefresh, "Unable to retrieve status information.")
//            appWidgetManager.updateAppWidget(appWidgetId, views)
//            return
//        }
        val isICEOrHybrid = carStatus.isPropulsionICEOrHybrid()
        val isPHEV = carStatus.isPropulsionPHEV()
        val isElectric = carStatus.isPropulsionElectric()
//        val isDiesel = carStatus.isPropulsionDiesel(carStatus.propulsion)

        views.setViewVisibility(R.id.lock_gasoline, if (isICEOrHybrid) View.VISIBLE else View.GONE)
        views.setViewVisibility(
            R.id.bottom_gasoline,
            if (isICEOrHybrid) View.VISIBLE else View.GONE
        )
        views.setViewVisibility(R.id.lock_electric, if (isICEOrHybrid) View.GONE else View.VISIBLE)
        views.setViewVisibility(
            R.id.bottom_electric,
            if (isICEOrHybrid) View.GONE else View.VISIBLE
        )
        views.setViewVisibility(R.id.plug, if (isICEOrHybrid) View.GONE else View.VISIBLE)
        setPHEVCallbacks(context, views, isPHEV, appWidgetId, "showGasoline")
//        setDieselCallbacks(context, views, isDiesel, appWidgetId, "showDEFLevel")
        setElectricCallbacks(context, views, isElectric, appWidgetId)

        // TODO: fix localization
        // Show last refresh, odometer, OTA status
        val timeFormat =
            if (false) Constants.LOCALTIMEFORMATUS else Constants.LOCALTIMEFORMAT
        drawLastRefresh(context, views, carStatus, timeFormat)
        drawOdometer(context, views, carStatus, distanceConversion, distanceUnits)
        //        drawOTAInfo(context, views, vehicleInfo, timeFormat);

        // Location
        if (PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(context.resources.getString(R.string.show_location_key), true)
        ) {
            views.setViewVisibility(R.id.location_container, View.VISIBLE)
            updateLocation(
                context,
                appWidgetManager,
                appWidgetId,
                views,
                carStatus.vehicle.vehicleLocation.latitude,
                carStatus.vehicle.vehicleLocation.longitude
            )
        } else {
            views.setViewVisibility(R.id.location_container, View.GONE)
        }

        // Ignition, alarm/sleep, plug icons
        drawIcons(views, carStatus)

        // Draw range and fuel/gas stuff
        val displayTime = true
        drawRangeFuel(
            context, views, carStatus, info, vehicleInfo,
            distanceConversion, distanceUnits, displayTime
        )

        // Tire pressures
        // Tire pressures
        updateTire(
            context, views,
            "", // carStatus.vehicle.vehicleStatus.tpms?.leftFrontTirePressure?.value,
            "", // carStatus.vehicle.vehicleStatus.tpms?.leftFrontTireStatus?.value,
            pressureUnits, pressureConversion, R.id.lt_ft_tire
        )
        updateTire(
            context, views,
            "", // carStatus.vehicle.vehicleStatus.tpms?.rightFrontTirePressure?.value,
            "", // carStatus.vehicle.vehicleStatus.tpms?.rightFrontTireStatus?.value,
            pressureUnits, pressureConversion, R.id.rt_ft_tire
        )
        updateTire(
            context, views,
            "", // carStatus.vehicle.vehicleStatus.tpms?.outerLeftRearTirePressure?.value,
            "", // carStatus.vehicle.vehicleStatus.tpms?.outerLeftRearTireStatus?.value,
            pressureUnits, pressureConversion, R.id.lt_rr_tire
        )
        updateTire(
            context, views,
            "", // carStatus.vehicle.vehicleStatus.tpms?.outerRightRearTirePressure?.value,
            "", // carStatus.vehicle.vehicleStatus.tpms?.outerRightRearTireStatus?.value,
            pressureUnits, pressureConversion, R.id.rt_rr_tire
        )

        // Get the right images to use for this vehicle
        val vehicleImages = getVehicle(vehicleInfo.modelId).horizontalDrawables

        // See if we should guess vehicle color
        if (scanImageForColor(context, vehicleInfo)) {
            info.setVehicle(vehicleInfo)
        }

        // TODO: re-enable if we're ever able to read the VIN again
        // If vehicle is a Mach-E First Edition, show mirrors in body color
//        if (isFirstEdition(context, vehicleInfo.vin!!)) {
//            vehicleImages[Vehicle.BODY_SECONDARY] = R.drawable.mache_secondary_no_mirrors_horz
//        }

        // Draw the vehicle image
        drawVehicleImage(context, views, carStatus, vehicleInfo.colorValue, null, vehicleImages)
        updateLinkedApps(context, views)

        // Instruct the widget manager to update the widget
        try {
            appWidgetManager.updateAppWidget(appWidgetId, views)
        } catch (_: NullPointerException) {
        }
    }

    override fun onAppWidgetOptionsChanged(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetId: Int,
        newOptions: Bundle
    ) {
        val views = RemoteViews(context.packageName, R.layout.widget_2x5)
        appWidgetManager.partiallyUpdateAppWidget(appWidgetId, views)
    }

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        LogFile.defineContext(context)
        CoroutineScope(Dispatchers.Main).launch {
            val info = getInfo(context)
            // There may be multiple widgets active, so update all of them
            for (appWidgetId in appWidgetIds) {
                updateAppWidget(context, appWidgetManager, appWidgetId, info)
            }
        }
    }

    override fun onEnabled(context: Context) {
        // Enter relevant functionality for when the first widget is created
    }

    override fun onDisabled(context: Context) {
        // Enter relevant functionality for when the last widget is disabled
    }

    override fun onReceive(context: Context, intent: Intent) {
        val action = intent.action
        when (action) {
            PHEVTOGGLE_CLICK -> {
                val mode = intent.getStringExtra("nextMode")
                val appWidgetManager = AppWidgetManager.getInstance(context)
                val views = RemoteViews(context.packageName, R.layout.widget_2x5)
                val nextMode: String
                if (mode == "showGasoline") {
                    nextMode = "showElectric"
                    views.setViewVisibility(R.id.bottom_electric, View.GONE)
                    views.setViewVisibility(R.id.bottom_gasoline, View.VISIBLE)
                } else {
                    nextMode = "showGasoline"
                    views.setViewVisibility(R.id.bottom_electric, View.VISIBLE)
                    views.setViewVisibility(R.id.bottom_gasoline, View.GONE)
                }
                val appWidgetId = intent.getIntExtra(APPWIDGETID, -1)
                setPHEVCallbacks(context, views, true, appWidgetId, nextMode)
                appWidgetManager.partiallyUpdateAppWidget(appWidgetId, views)
            }
            DIESELTOGGLE_CLICK -> {
                val mode = intent.getStringExtra("nextMode")
                val appWidgetManager = AppWidgetManager.getInstance(context)
                val views = RemoteViews(context.packageName, R.layout.widget_2x5)
                val nextMode: String
                when (mode) {
                    "showDEFLevel" -> {
                        nextMode = "showDEFRange"
                        views.setViewVisibility(R.id.LVBDisplay, View.GONE)
                        views.setViewVisibility(R.id.DEFLevel, View.VISIBLE)
                        views.setViewVisibility(R.id.DEFRange, View.GONE)
                    }
                    "showDEFRange" -> {
                        nextMode = "showLVBVoltage"
                        views.setViewVisibility(R.id.LVBDisplay, View.GONE)
                        views.setViewVisibility(R.id.DEFLevel, View.GONE)
                        views.setViewVisibility(R.id.DEFRange, View.VISIBLE)
                    }
                    else -> {
                        nextMode = "showDEFLevel"
                        views.setViewVisibility(R.id.LVBDisplay, View.VISIBLE)
                        views.setViewVisibility(R.id.DEFLevel, View.GONE)
                        views.setViewVisibility(R.id.DEFRange, View.GONE)
                    }
                }
                val appWidgetId = intent.getIntExtra(APPWIDGETID, -1)
                setDieselCallbacks(context, views, true, appWidgetId, nextMode)
                appWidgetManager.partiallyUpdateAppWidget(appWidgetId, views)
            }
            else -> {
                super.onReceive(context, intent)
            }
        }
    }
}
